keanu-project/src/main/java/io/improbable/keanu/network/BayesianNetwork.java

```diff
@@ -214,15 +222,49 @@ public void incrementIndentation() {
         indentation++;
     }
 
-    public void save(NetworkWriter networkWriter) throws IOException {
+    public void save(NetworkWriter networkWriter) {
         for (Vertex vertex : TopologicalSort.sort(vertices)) {
             vertex.save(networkWriter);
         }
     }
 
-    public void saveValues(NetworkWriter networkWriter) throws IOException {
+    public void saveValues(NetworkWriter networkWriter) {
         for (Vertex vertex : vertices) {
             vertex.saveValue(networkWriter);
         }
     }
+
+    /**
+     * Method for traversing a graph and returning a subgraph of vertices within the given degree of the specified vertex.
+     *
+     * @param vertex vertex that the subgraph will be centered around
+     * @param degree degree of connections from the vertex to be included in the subgraph
+     * @return a set of vertices within the specified degree from the given vertex
+     */
+    public Set<Vertex> getSubgraph(Vertex vertex, int degree) {
+
+        Set<Vertex> subgraphVertices = new HashSet<>();
+        Set<Vertex> verticesToProcessNow = new HashSet<>();
+        verticesToProcessNow.add(vertex);
+
+        int iterationIndex = 0;
+        while (iterationIndex <= degree && !verticesToProcessNow.isEmpty()) {
+
+            Set<Vertex> verticesToProcessNext = new HashSet<>();
+            subgraphVertices.addAll(verticesToProcessNow);
+
+            for (Vertex v : verticesToProcessNow) {
+                Stream<Vertex> connectedVertices = Stream.concat(v.getParents().stream(), v.getChildren().stream());
+                connectedVertices.forEach(connectedVertex -> {
+                    if (!subgraphVertices.contains(connectedVertex)) {
```

Comment:
Streams have a .filter() method which would simplify the look of this code a bit (ie you wouldn't need the "if")

Reply:
Changed it to use filter.